### PR TITLE
feat(claude): discovery rigor, finding categorization, and /polish

### PR DESCRIPTION
**Pull request by inkatze**
## Summary
- Adds **Discovery Rigor**, **Refactor Instinct**, and **Finding Categorization** as canonical sections in the global `CLAUDE.md`. Discovery Rigor mandates a 9-lens checklist with a lens-coverage table, parallel sub-agent fan-out for non-trivial diffs, and a self-critique pass. Refactor Instinct splits behavior between implementation mode (low bar, clean as you go) and review mode (high bar, only flag if this PR worsens structure). Finding Categorization defines four bright-line conditions for Auto-applicable plus disqualifiers (security, migrations, CI, lockfiles, secrets).
- Adds **`/polish`**: an autonomous loop of `/self-review`'s discovery + validation that applies only Auto-applicable items (tool-grounded, mechanical, no behavior change, validation passes converged). Local-only (no push, no PR), cap = 10, hands off the moment any Needs-human-attention item or safety condition fires.
- Adds **tool-discovery SessionStart hook** (`roles/osx/files/claude/scripts/tool-discovery.sh`): scans cwd for known linter / formatter / type-checker / hook-manager / CI config files and emits a markdown summary as `additionalContext`, so Discovery Rigor and Refactor Instinct decisions have grounded tooling input without grepping. Bash 3.2 compatible.
- Updates `/self-review` and `/code-review` to use parallel `Explore` sub-agent fan-out per lens, and switches all four review-skill outputs (`/self-review`, `/code-review`, `/copilot-review` adjacent findings, `/peer-review`) to two-table layout split by Auto-applicable vs Needs human attention. `/copilot-pairing` is intentionally not changed; it inherits step 4's fan-out via the existing `/copilot-review` reference and its autonomous-loop semantics are out of scope.
- Adds `.github/copilot-instructions.md` so Copilot's own reviews of this dotfiles repo follow the same single-pass exhaustiveness, refactor-instinct split, and tool-grounded rule. Uses this PR as the first live test.

## Test plan
- [ ] After the next Ansible run, `readlink ~/.claude/commands/polish.md` resolves into the tracked source and `~/.claude/scripts/tool-discovery.sh` is executable.
- [ ] Fresh Claude session in this repo emits a `## Project tooling (auto-detected)` block as additionalContext listing lefthook, mise, yamllint, ansible-lint, and `.github/workflows`.
- [ ] Fresh Claude session in a non-git directory: tool-discovery hook is silent (no JSON output, no error).
- [ ] `/self-review` against a small diff prints the canonical lens-coverage table followed by both Auto-applicable and Needs-human-attention tables (`none` rows when empty).
- [ ] `/polish` refuses to start when `git status --porcelain` is non-empty (Dirty working tree stop condition).
- [ ] `/polish` on a branch with one Auto-applicable item (e.g., a deliberately introduced unused import the linter catches) applies the fix, commits as `chore(polish): iter 1, ...`, then loops once more and exits cleanly with no findings.
- [ ] `/polish` on a branch that mixes one Auto-applicable and one Needs-human-attention finding applies the Auto-applicable item in iter 1, then in iter 2 surfaces the Needs-human-attention item and exits via the **Human attention required** stop condition without further commits.
- [ ] Copilot's review on this PR follows the new `.github/copilot-instructions.md` (lens-walk, tool-grounded, refactor-instinct review-mode bar). Verifiable by inspecting the threads Copilot files.

### CODE REVIEW

File: roles/osx/files/claude/scripts/tool-discovery.sh [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+# TypeScript / JavaScript
+[ -f tsconfig.json ] && add "tsc (\`tsc --noEmit\`)"
+if [ -f .eslintrc.json ] || [ -f .eslintrc.js ] || [ -f .eslintrc.cjs ] || [ -f .eslintrc.yml ] || [ -f eslint.config.js ] || [ -f eslint.config.mjs ] || [ -f eslint.config.cjs ]; then
+    add "eslint (\`eslint .\`)"
+fi
+[ -f biome.json ] || [ -f biome.jsonc ] && add "biome (\`biome check .\`)"
+if [ -f .prettierrc ] || [ -f .prettierrc.json ] || [ -f .prettierrc.js ] || [ -f .prettierrc.yml ] || [ -f prettier.config.js ] || [ -f prettier.config.cjs ]; then
+    add "prettier (\`prettier --check .\`)"
+fi
+[ -f knip.json ] || [ -f .knip.json ] || [ -f knip.config.ts ] && add "knip (dead-code: \`knip\`)"
```
[the remark this answers is not in the document]
Comment (inkatze) [reply roles/osx/files/claude/scripts/tool-discovery.sh:84]: False positive. `&&` and `||` between simple commands have equal precedence and are **left-associative** in bash, so `[ -f a ] || [ -f b ] && add ...` parses as `([ -f a ] || [ -f b ]) && add ...`, and `add` runs whenever **any** test is true (not only the last one).

Reproduced in bash 5.3 against this exact pattern:

| case | files present | `add` ran? |
| --- | --- | --- |
| A | `mypy.ini` only | yes |
| B | `.mypy.ini` only | yes |
| C | neither | no |

Same outcomes for the `biome.json | biome.jsonc` chain on line 76 and the `build.gradle | build.gradle.kts | pom.xml` triple chain on line 100.

Three passes converged:

- Pass 1 (reproduction): traced exit codes across each branch in bash 5.3.
- Pass 2 (orthogonal): the bash manual section "Lists of Commands" states *"AND and OR lists are executed with left associativity"*, which is exactly the parse we want.
- Pass 3 (outside-in): same wording in the GNU bash reference manual.

Not changing this pattern.

File: roles/osx/files/claude/scripts/tool-discovery.sh [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+if [ -f package.json ]; then
```
[the remark this answers is not in the document]
Comment (inkatze) [reply roles/osx/files/claude/scripts/tool-discovery.sh:85]: False positive. `&&` and `||` between simple commands have equal precedence and are **left-associative** in bash, so `[ -f a ] || [ -f b ] && add ...` parses as `([ -f a ] || [ -f b ]) && add ...`, and `add` runs whenever **any** test is true (not only the last one).

Reproduced in bash 5.3 against this exact pattern:

| case | files present | `add` ran? |
| --- | --- | --- |
| A | `mypy.ini` only | yes |
| B | `.mypy.ini` only | yes |
| C | neither | no |

Same outcomes for the `biome.json | biome.jsonc` chain on line 76 and the `build.gradle | build.gradle.kts | pom.xml` triple chain on line 100.

Three passes converged:

- Pass 1 (reproduction): traced exit codes across each branch in bash 5.3.
- Pass 2 (orthogonal): the bash manual section "Lists of Commands" states *"AND and OR lists are executed with left associativity"*, which is exactly the parse we want.
- Pass 3 (outside-in): same wording in the GNU bash reference manual.

Not changing this pattern.

File: roles/osx/files/claude/scripts/tool-discovery.sh [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+
+# Java / Kotlin
+[ -f build.gradle ] || [ -f build.gradle.kts ] || [ -f pom.xml ] && add "JVM build (\`./gradlew check\` or \`mvn verify\`)"
+
+# Shell
+[ -f .shellcheckrc ] && add "shellcheck (\`shellcheck **/*.sh\`)"
+
+# YAML / Ansible (this dotfiles repo uses these)
+[ -f .yamllint ] || [ -f .yamllint.yml ] || [ -f .yamllint.yaml ] && add "yamllint (\`yamllint .\`)"
+[ -f .ansible-lint ] || [ -f .ansible-lint.yml ] || [ -f ansible-lint.yml ] && add "ansible-lint (\`ansible-lint\`)"
```
[the remark this answers is not in the document]
Comment (inkatze) [reply roles/osx/files/claude/scripts/tool-discovery.sh]: False positive. `&&` and `||` between simple commands have equal precedence and are **left-associative** in bash, so `[ -f a ] || [ -f b ] && add ...` parses as `([ -f a ] || [ -f b ]) && add ...`, and `add` runs whenever **any** test is true (not only the last one).

Reproduced in bash 5.3 against this exact pattern:

| case | files present | `add` ran? |
| --- | --- | --- |
| A | `mypy.ini` only | yes |
| B | `.mypy.ini` only | yes |
| C | neither | no |

Same outcomes for the `biome.json | biome.jsonc` chain on line 76 and the `build.gradle | build.gradle.kts | pom.xml` triple chain on line 100.

Three passes converged:

- Pass 1 (reproduction): traced exit codes across each branch in bash 5.3.
- Pass 2 (orthogonal): the bash manual section "Lists of Commands" states *"AND and OR lists are executed with left associativity"*, which is exactly the parse we want.
- Pass 3 (outside-in): same wording in the GNU bash reference manual.

Not changing this pattern.

File: roles/osx/files/claude/scripts/tool-discovery.sh [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+#
+# Output: a small JSON blob with `additionalContext` if any tooling is
+# detected, or no output (silent) if nothing is found. Uses bash 3.2 +
+# `set -u` to stay compatible with the macOS default bash.
+
+set -u
+
+cwd="${CLAUDE_PROJECT_DIR:-$PWD}"
+[ -d "$cwd" ] || exit 0
+cd "$cwd" 2>/dev/null || exit 0
```
[the remark this answers is not in the document]
Comment (inkatze) [reply roles/osx/files/claude/scripts/tool-discovery.sh:19]: Addressed in e8b6dac. Added an early exit when `git rev-parse --is-inside-work-tree` fails (or `git` is unavailable), so the hook stays silent in non-git directories. Matches the PR test plan ("non-git directory: tool-discovery hook is silent").

Reproduced before the fix: a non-git directory containing only a `Makefile` produced 323 bytes of `additionalContext`. After the fix: zero bytes (silent exit).

File: roles/osx/files/claude/scripts/tool-discovery.sh [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+
+Use these for tool-grounded discovery and refactor decisions (per CLAUDE.md \`Discovery Rigor\` and \`Refactor Instinct\`). Tool output is grounded; vibes are not.
+"
+for ((i=0; i<total; i++)); do
+    summary="${summary}
+- ${names[$i]}"
+done
+
+printf '{"hookSpecificOutput":{"hookEventName":"SessionStart","additionalContext":%s}}\n' \
+    "$(printf '%s' "$summary" | jq -Rs .)"
```
[the remark this answers is not in the document]
Comment (inkatze) [reply roles/osx/files/claude/scripts/tool-discovery.sh:151]: Addressed in e8b6dac. Added `command -v jq >/dev/null 2>&1 || exit 0` ahead of the `printf` so the hook stays silent when `jq` is missing rather than emitting an invalid `additionalContext` payload.

Reproduced before the fix by running the script under a `PATH` that excluded `jq`: output was

```
{"hookSpecificOutput":{"hookEventName":"SessionStart","additionalContext":}}
```

which fails JSON parsing. After the fix, the same scenario produces zero output and exits 0.

File: roles/osx/files/claude/scripts/tool-discovery.sh [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+fi
+
+# Erlang
+[ -f rebar.config ] && add "rebar3 (\`rebar3 dialyzer\`)"
+
+# Java / Kotlin
+[ -f build.gradle ] || [ -f build.gradle.kts ] || [ -f pom.xml ] && add "JVM build (\`./gradlew check\` or \`mvn verify\`)"
+
+# Shell
+[ -f .shellcheckrc ] && add "shellcheck (\`shellcheck **/*.sh\`)"
```
[the remark this answers is not in the document]
Comment (inkatze) [reply roles/osx/files/claude/scripts/tool-discovery.sh]: Addressed in e8b6dac. Replaced the suggested invocation with `git ls-files '*.sh' | xargs shellcheck`, which works under macOS default bash 3.2 without needing `shopt -s globstar` and stays scoped to tracked files in a git repo (which is now a precondition for the hook running at all, per the thread on line 18).

Verified the prior form's failure mode: in a fresh bash with `globstar` off, `**/*.sh` only matched files in subdirectories, silently missing top-level scripts.

File: .github/copilot-instructions.md [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+each finding before posting:
+
+1. **Reproduce or trace.** For runtime claims, mentally execute with concrete
+   inputs or point to the exact failing case. If you cannot, downgrade or drop.
+2. **Orthogonal check.** Look at callers, sibling implementations, existing
+   tests, project conventions. If the pattern is intentional and consistent
+   elsewhere, it is probably not a bug.
+3. **Outside-in check.** Consult `git log` / `git blame` for why-it-is-the-way-
+   it-is. Check `CLAUDE.md`, `specs/README.md`, and per-directory docs for
+   documented conventions before flagging a "violation."
```
[the remark this answers is not in the document]
Comment (inkatze) [reply .github/copilot-instructions.md]: Addressed in e8b6dac. Reworded the bullet so the phrase no longer splits across a line with a trailing hyphen. New text:

> Consult `git log` / `git blame` for context on why the code is shaped the way it is.

This avoids the rendering artifact while keeping the original meaning of the rule.

File: roles/osx/files/claude/scripts/tool-discovery.sh [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+git rev-parse --is-inside-work-tree >/dev/null 2>&1 || exit 0
+
+names=()
+add() { names+=("$1"); }
+
+# Hook managers / multi-tool runners
+if [ -f lefthook.yml ] || [ -f .lefthook.yml ] || [ -f lefthook.yaml ]; then
+    add "lefthook (run pre-commit hooks: \`lefthook run pre-commit\`)"
+fi
+if [ -f .pre-commit-config.yaml ]; then
```
[the remark this answers is not in the document]
Comment (inkatze) [reply roles/osx/files/claude/scripts/tool-discovery.sh]: Partially addressed in 6e94d03. Added `.lefthook.yaml` to the lefthook check (lefthook does load that variant per its own docs).

Did **not** add `.pre-commit-config.yml`: pre-commit's source declares `CONFIG_FILE = '.pre-commit-config.yaml'` as the only supported filename ([pre-commit/pre-commit constants](https://github.com/pre-commit/pre-commit/blob/main/pre_commit/constants.py)), so detecting `.pre-commit-config.yml` would surface a "pre-commit detected" line for a file that pre-commit itself ignores. False-positive detection in the additionalContext is worse than the false negative this thread flagged.

Verified: a fresh git repo containing `.lefthook.yaml` now triggers the lefthook line.

File: roles/osx/files/claude/scripts/tool-discovery.sh [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+        add "Python (pyproject.toml present; check for ruff/mypy/black/etc.)"
+    fi
+fi
+[ -f mypy.ini ] || [ -f .mypy.ini ] && add "mypy (\`mypy .\`)"
+[ -f pyrightconfig.json ] && add "pyright (\`pyright\`)"
+[ -f .ruff.toml ] || [ -f ruff.toml ] && add "ruff (\`ruff check .\`)"
+
+# TypeScript / JavaScript
+[ -f tsconfig.json ] && add "tsc (\`tsc --noEmit\`)"
+if [ -f .eslintrc.json ] || [ -f .eslintrc.js ] || [ -f .eslintrc.cjs ] || [ -f .eslintrc.yml ] || [ -f eslint.config.js ] || [ -f eslint.config.mjs ] || [ -f eslint.config.cjs ]; then
```
[the remark this answers is not in the document]
Comment (inkatze) [reply roles/osx/files/claude/scripts/tool-discovery.sh]: Addressed in 6e94d03. Added `.eslintrc.yaml` to the eslint chain. Verified: a fresh git repo containing only `.eslintrc.yaml` now triggers the eslint detection line.

File: roles/osx/files/claude/scripts/tool-discovery.sh [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+
+# Java / Kotlin
+[ -f build.gradle ] || [ -f build.gradle.kts ] || [ -f pom.xml ] && add "JVM build (\`./gradlew check\` or \`mvn verify\`)"
+
+# Shell
+[ -f .shellcheckrc ] && add "shellcheck (\`git ls-files '*.sh' | xargs shellcheck\`)"
+
+# YAML / Ansible (this dotfiles repo uses these)
+[ -f .yamllint ] || [ -f .yamllint.yml ] || [ -f .yamllint.yaml ] && add "yamllint (\`yamllint .\`)"
+[ -f .ansible-lint ] || [ -f .ansible-lint.yml ] || [ -f ansible-lint.yml ] && add "ansible-lint (\`ansible-lint\`)"
```
[the remark this answers is not in the document]
Comment (inkatze) [reply roles/osx/files/claude/scripts/tool-discovery.sh]: Addressed in 6e94d03. Added both `.ansible-lint.yaml` and `ansible-lint.yaml` to the ansible-lint chain. Verified each variant separately triggers the ansible-lint detection line.

File: roles/osx/files/claude/scripts/tool-discovery.sh [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+fi
+
+# Erlang
+[ -f rebar.config ] && add "rebar3 (\`rebar3 dialyzer\`)"
+
+# Java / Kotlin
+[ -f build.gradle ] || [ -f build.gradle.kts ] || [ -f pom.xml ] && add "JVM build (\`./gradlew check\` or \`mvn verify\`)"
+
+# Shell
+[ -f .shellcheckrc ] && add "shellcheck (\`git ls-files '*.sh' | xargs shellcheck\`)"
```
[the remark this answers is not in the document]
Comment (inkatze) [reply roles/osx/files/claude/scripts/tool-discovery.sh]: Addressed in 6e94d03. Switched the suggested invocation to `git ls-files -z '*.sh' | xargs -0 shellcheck`, which is NUL-delimited end-to-end and stays correct on tracked `.sh` paths containing whitespace.

File: roles/osx/files/claude/scripts/tool-discovery.sh [hunk 12]
```diff
@@ -0,0 +1,149 @@
...
+[ -f pyrightconfig.json ] && add "pyright (\`pyright\`)"
+[ -f .ruff.toml ] || [ -f ruff.toml ] && add "ruff (\`ruff check .\`)"
+
+# TypeScript / JavaScript
+[ -f tsconfig.json ] && add "tsc (\`tsc --noEmit\`)"
+if [ -f .eslintrc.json ] || [ -f .eslintrc.js ] || [ -f .eslintrc.cjs ] || [ -f .eslintrc.yml ] || [ -f eslint.config.js ] || [ -f eslint.config.mjs ] || [ -f eslint.config.cjs ]; then
+    add "eslint (\`eslint .\`)"
+fi
+[ -f biome.json ] || [ -f biome.jsonc ] && add "biome (\`biome check .\`)"
+if [ -f .prettierrc ] || [ -f .prettierrc.json ] || [ -f .prettierrc.js ] || [ -f .prettierrc.yml ] || [ -f prettier.config.js ] || [ -f prettier.config.cjs ]; then
```
[the remark this answers is not in the document]
Comment (inkatze) [reply roles/osx/files/claude/scripts/tool-discovery.sh]: Addressed in 6e94d03. Added `.prettierrc.yaml` to the prettier chain. Verified: a fresh git repo containing only `.prettierrc.yaml` now triggers the prettier detection line.

File: roles/osx/files/claude/scripts/tool-discovery.sh [hunk 13]
```diff
@@ -0,0 +1,149 @@
...
+# Wired from ~/.claude/settings.json.
+#
+# Purpose: Discovery Rigor and Refactor Instinct in CLAUDE.md want the agent
+# to ground decisions in what the project actually runs (linters, formatters,
+# type checkers, etc.) rather than vibes. This hook does the cwd scan once at
+# session start so the agent sees a short list without having to grep around.
+#
+# Output: a small JSON blob with `additionalContext` if any tooling is
+# detected, or no output (silent) if nothing is found. Uses bash 3.2 +
+# `set -u` to stay compatible with the macOS default bash.
```
[the remark this answers is not in the document]
Comment (inkatze) [reply roles/osx/files/claude/scripts/tool-discovery.sh]: Addressed in 6ce0298. Updated the header comment to enumerate all silent-exit conditions:

> Silent no-op (no output, exit 0) when any of: nothing detected, the cwd is outside a git work tree, or `jq` is unavailable.

File: CLAUDE.md [hunk 14]
```diff
@@ -72,6 +72,20 @@ with no sandboxing, so opening Claude in an untrusted checkout executes
 whatever that script contains. Same trust model as `mise trust`: inspect the
 script before opening a repo you did not author.
 
+### Tool-discovery hook
+
+`roles/osx/files/claude/scripts/tool-discovery.sh` runs on `SessionStart`
+alongside the worktree bootstrap. It scans the cwd for known config files
+(linters, formatters, type checkers, hook managers, CI workflows) and emits
+a markdown summary as `additionalContext`, so the agent sees what the project
+ships without grepping. Silent no-op when nothing is detected. Read-only,
```
[the remark this answers is not in the document]
Comment (inkatze) [reply CLAUDE.md]: Addressed in 6ce0298. Reworded the tool-discovery section in `CLAUDE.md` so the silent-exit conditions are listed and a missing summary is explicitly flagged as ambiguous:

> Silent no-op (exit 0, no output) when any of: nothing is detected, the cwd is outside a git work tree, or `jq` is unavailable; a missing summary therefore does not necessarily mean "no tooling found".

File: roles/osx/files/claude/commands/polish.md [hunk 15]
```diff
@@ -0,0 +1,131 @@
+Iterate `/self-review` autonomously, applying only Auto-applicable items, until none remain. Hand control back the moment anything Needs human attention or any safety condition fires.
+
+Same Discovery + Validation rigor as `/self-review`, executed on autopilot, with hard stop conditions baked in for safety.
+
+## When to use
+
+You are nearly done with a branch and want to drain the trivial, tool-grounded fixes (linter rule violations, formatter output, type-checker errors, missing imports, unused variables, typos in comments, inferable type annotations) before opening a PR. Polish auto-applies items that meet the strict Auto-applicable definition in CLAUDE.md `Finding Categorization` and stops the moment anything needs your judgment.
+
+Polish does **not** create a PR. Run `/self-review` after Polish hands off, once you have addressed the Needs human attention items.
```
[the remark this answers is not in the document]
Comment (inkatze) [reply roles/osx/files/claude/commands/polish.md]: Addressed in 6ce0298, after a short alignment pause to confirm intent. Two valid interpretations existed:

- **A.** Stop on the first Needs-human-attention item (matched the opening prose + workflow list).
- **B.** Drain Auto-applicable items each iteration, surface Needs-human-attention items in the iteration tables, only hand off when Auto-applicable is empty or another safety condition fires (matched step (c) and the PR test plan).

The repo owner picked **B**. The commit reworks the opening line of `polish.md` (and the workflow-list bullet in `roles/osx/files/CLAUDE.md`) to describe Behavior B explicitly:

> Hand control back when no Auto-applicable items are left to drain (surfacing any Needs human attention items in that final table) or any safety condition fires.

Step (c)'s loop logic was already correct under Behavior B, so it stays as-is. polish.md:9, :22, :54, :55, :113 were re-checked and are all consistent.

File: roles/osx/files/claude/commands/polish.md [hunk 16]
```diff
@@ -0,0 +1,131 @@
...
+
+Be more conservative than `/self-review` because nobody is checking the categorization in real time. **When in doubt, route to Needs human attention.** False negatives (a real Auto-applicable item routed to human) are cheap, costing one extra iteration. False positives (a judgment item auto-applied) silently corrupt the branch.
+
+### c. Decide loop fate
+
+Branch on the bucket counts:
+
+- **Both buckets empty.** Success. Exit the loop. Print the final summary noting "no findings remain". Do not commit.
+- **Auto-applicable empty, Needs human attention non-empty.** Stop. Trigger **Human attention required** stop condition. Print the latest tables and hand control back. Do not push, do not commit, do not auto-apply anything.
+- **Auto-applicable non-empty, regardless of Needs human attention count.** Proceed to step (d). Items in Needs human attention will be re-evaluated next iteration; the user can address them after Polish hands off.
```
[the remark this answers is not in the document]
Comment (inkatze) [reply roles/osx/files/claude/commands/polish.md:55]: Addressed in 6ce0298. Step (c) is the canonical statement of Polish's handoff semantics; the contradicting prose was the opening line of `polish.md` (and the matching bullet in `roles/osx/files/CLAUDE.md`). Both have been reworded to Behavior B (drain Auto-applicable items, hand off when none remain). See the response on the related thread on the opening summary for the full picture.

File: roles/osx/files/CLAUDE.md [hunk 17]
```diff
@@ -114,13 +114,98 @@ For any fix, validate the solution with at least two independent test angles, th
...
-3. **Copilot pairing** (`/copilot-pairing`): Same rigor as `/copilot-review`, but loops autonomously: address Copilot's threads, push, re-request review, wait for Copilot's response, repeat until Copilot has no new comments. Hard stop conditions (ambiguity, scope creep, test failure, security-sensitive code, loop detection, iteration cap of 10) hand control back to the human.
-4. **Peer review** (`/peer-review`): Address unresolved peer review threads on the current PR. Same validation process as Copilot review, but responses must sound natural, human, and match the user's communication style.
+2. **Polish** (`/polish`): Autonomous loop of `/self-review`'s discovery + validation, applying only Auto-applicable items (tool-grounded, mechanical, no behavior change, validation passes converged). Hands off the moment any Needs human attention item appears or any safety condition fires. Local-only: no push, no PR. Use as a finishing pass before `/self-review` opens the PR.
```
[the remark this answers is not in the document]
Comment (inkatze) [reply roles/osx/files/CLAUDE.md]: Addressed in 6ce0298. The workflow-list bullet here was the prose mismatch; step (c) of `polish.md` already implemented the intended Behavior B (drain Auto-applicable, hand off when none remain). Bullet now reads:

> Each iteration drains the Auto-applicable bucket and lets Needs human attention items accumulate; the loop only hands off when no Auto-applicable items remain (surfacing the Needs human attention items at that point) or any safety condition fires.

File: roles/osx/files/claude/commands/polish.md [hunk 18]
```diff
@@ -0,0 +1,131 @@
+Iterate `/self-review` autonomously, applying only Auto-applicable items, until none remain. Hand control back when no Auto-applicable items are left to drain (surfacing any Needs human attention items in that final table) or any safety condition fires.
+
+Same Discovery + Validation rigor as `/self-review`, executed on autopilot, with hard stop conditions baked in for safety.
+
+## When to use
+
+You are nearly done with a branch and want to drain the trivial, tool-grounded fixes (linter rule violations, formatter output, type-checker errors, missing imports, unused variables, typos in comments, inferable type annotations) before opening a PR. Polish auto-applies items that meet the strict Auto-applicable definition in CLAUDE.md `Finding Categorization` and stops the moment anything needs your judgment.
```
[the remark this answers is not in the document]
Comment (inkatze) [reply roles/osx/files/claude/commands/polish.md]: Addressed in 32742a2. The "When to use" sentence was a Behavior A leftover that iter 3's reword (opening line of `polish.md` + workflow-list bullet) did not pick up. Now matches Behavior B end-to-end:

> Polish auto-applies items that meet the strict Auto-applicable definition in CLAUDE.md `Finding Categorization`; once the Auto-applicable bucket is empty (or any safety condition fires), it surfaces any Needs human attention items in the final iteration table and hands off.

Re-grepped the rest of `polish.md` and the materialized `CLAUDE.md` for any other "stop the moment / hands off the moment" prose; none remain.

File: .github/copilot-instructions.md [hunk 19]
```diff
@@ -0,0 +1,148 @@
...
+  deliberate so mise-managed runtimes are picked up.
+- **Ansible materialization model.** Files under `roles/osx/files/claude/` are
+  the source of truth. The runtime path `~/.claude/...` is materialized via
+  symlink (most files) or `jq` merge (`settings.json`). Do not suggest editing
+  the runtime path.
+- **MCP secret handling.** `scripts/claude-mcp-sync-github.sh` reads the
+  GitHub PAT from 1Password, writes via atomic `jq` temp+rename, and scopes
+  `GITHUB_PAT` to two `jq` invocations. Argv exposure and same-user env
+  inspection are documented and accepted; do not re-flag them. The strict-
+  fail on a locked vault is **deliberate** (a silent skip would let stale PATs
```
[the remark this answers is not in the document]
Comment (inkatze) [reply .github/copilot-instructions.md]: Addressed in b58cbe6. Reflowed the bullet so `strict-fail` stays on a single line:

> The strict-fail on a locked vault is **deliberate**...

Also did a sweep across the other files in this PR (`polish.md`, `roles/osx/files/CLAUDE.md`, `CLAUDE.md`, `tool-discovery.sh`) for any remaining hyphen-at-EOL line breaks; none found.

File: roles/osx/files/claude/scripts/tool-discovery.sh [hunk 20]
```diff
@@ -0,0 +1,150 @@
...
+    add "lefthook (run pre-commit hooks: \`lefthook run pre-commit\`)"
+fi
+if [ -f .pre-commit-config.yaml ]; then
+    add "pre-commit (\`pre-commit run --all-files\`)"
+fi
+if [ -f Makefile ] || [ -f makefile ] || [ -f GNUmakefile ]; then
+    add "Makefile (\`make\` targets; try \`make help\`)"
+fi
+if [ -f mise.toml ] || [ -f .mise.toml ] || [ -f .tool-versions ]; then
+    add "mise tasks (\`mise tasks\` to list)"
```
[the remark this answers is not in the document]
Comment (inkatze) [reply roles/osx/files/claude/scripts/tool-discovery.sh]: Addressed in a745491. Split the check so `.tool-versions` no longer suggests `mise tasks`:

```
if [ -f mise.toml ] || [ -f .mise.toml ]; then
    add "mise tasks (\`mise tasks\` to list)"
fi
if [ -f .tool-versions ]; then
    add ".tool-versions (runtime version pins; \`mise current\` to inspect)"
fi
```

Verified: a fresh git repo with only `.tool-versions` produces the runtime-pins line and no `mise tasks` line; `mise.toml` and `.mise.toml` still produce the tasks line.

File: roles/osx/files/claude/commands/peer-review.md [hunk 21]
```diff
@@ -69,13 +69,31 @@ For each remaining thread, apply the canonical rigor in CLAUDE.md `Validation Ri
...
+|---|---|---|---|---|---|---|---|---|
+
+- **Rule cited**: the linter / type-checker / formatter rule that confirms the reviewer's concern.
+- **Draft response**: short, polite. "Done in `<sha>`" or "Good catch, fixed in `<sha>`" is usually enough.
+
+**Needs human attention** (preference, design, refactor, ambiguity, low-confidence, anything that needs a real reply):
+
+| # | Thread ID | File:Line | Reviewer's concern | Our assessment | Classification | Confidence | Validation passes | Recommendation | Draft response |
+|---|---|---|---|---|---|---|---|---|---|
+
```
[the remark this answers is not in the document]
Comment (inkatze) [reply roles/osx/files/claude/commands/peer-review.md]: Addressed in a745491. Renamed `Our assessment` to `What we found` (matching `/copilot-review`'s convention) and added explicit definitions in both table legends:

- **Auto-applicable**: "a one-line, plain-prose verdict from our investigation. For Auto-applicable items this is usually 'tool confirmed, fix is mechanical' or similar."
- **Needs human attention**: "a one-line, plain-prose verdict from our investigation (the rationale that supports the `Classification` bucket)."

So `Classification` stays as the typed bucket label and `What we found` is the rationale that justifies it; they no longer collide.

File: roles/osx/files/claude/commands/copilot-pairing.md [hunk 22]
```diff
@@ -105,7 +105,18 @@ Order matters: land the code first, then talk about it. If we replied/resolved b
...
-3. **Reply to threads** using `/copilot-review` step 8 mutations (multi-line GraphQL, body via temp file). **Use `addPullRequestReviewThreadReply` only**; see the DO-NOT-USE callout in `/copilot-review` step 8 about `addPullRequestReviewComment`. Reply body varies by classification, since this iteration may have a mix:
+3. **Reply to threads** using `/copilot-review` step 8 mutations. **Use `addPullRequestReviewThreadReply` only**; see the DO-NOT-USE callout in `/copilot-review` step 8 about `addPullRequestReviewComment`. **Body construction: inline bash heredoc, not temp file.** Build the body inside the same `Bash` invocation that runs the GraphQL mutation:
+
+   ```bash
+   body=$(cat <<'EOF'
+   Addressed in <sha>. <short prose>
+   EOF
+   )
+   gh api graphql -f query='mutation($threadId: ID!, $body: String!) { addPullRequestReviewThreadReply(input: {pullRequestReviewThreadId: $threadId, body: $body}) { comment { id } } }' \
+     -f threadId='<thread_id>' -f body="$body"
```
[the remark this answers is not in the document]
Comment (inkatze) [reply roles/osx/files/claude/commands/copilot-pairing.md]: Addressed in 4c3ae08: rewrote the example as a multi-line `-f query='\n  mutation...\n'` block, matching `/copilot-review` step 8 and `/peer-review` step 5 verbatim. The single-line form was the only outlier in the skills tree (grep across `roles/osx/files/claude/commands/*.md` found this one occurrence).

File: roles/osx/files/CLAUDE.md [hunk 23]
```diff
@@ -112,15 +112,100 @@ For any fix, validate the solution with at least two independent test angles, th
...
 3. **Edge / integration / manual.** When relevant: boundary cases (null, empty, max size, concurrency), integration or smoke tests, manual exercise of the user-facing flow.
 
-For non-testable changes (docs, comments, formatting, pure renames, type-only adjustments): substitute review angles. Re-read the diff, read it from the perspective of each caller, and grep the repo for places the change could silently break. Note in the reply why a test was not added.
+For non-testable changes (docs, comments, formatting, pure renames, type-only adjustments): substitute review angles. Re-read the diff, read it from the perspective of each caller, and grep the repo for places the change could silently break. **For contract rewords (a doc rule expressed in several places, a behavior summary that recurs in workflow lists, a rename touching prose as well as code identifiers): grep the affected files for the surface patterns of the rule before declaring alignment, not only the lines a thread points at — otherwise stragglers surface as new threads in the next review cycle.** Note in the reply why a test was not added.
```
[the remark this answers is not in the document]
Comment (inkatze) [reply roles/osx/files/CLAUDE.md]: Addressed in 4c3ae08: replaced the em-dash with a sentence break (". Otherwise stragglers ...") to match the Writing Style rule. grep " — " across the file confirmed this was the only em-dash in `roles/osx/files/CLAUDE.md`.
